### PR TITLE
feat: Output of all information required by federated results

### PR DESCRIPTION
**Pull request by kobayu858**
###SUMMARY
- Output of all information required by federated results.
- DAG Information
 - Total utilization
 - Critical path length, deadline, and total workload for each DAG
- Processor Information
 - Total number of cores
- Result details
 - Scheduling success or failure
 - (in case of scheduling success)
  - Number of cores used by high utilization tasks
  - Number of cores used by underutilized tasks
 - (in case of scheduling failure)
 - Reason
 - Number of underutilized cores

### CODE REVIEW

File: lib/src/output_log.rs [hunk 1]
```diff
@@ -20,6 +21,11 @@ struct DAGInfo {
     volume: f32,
 }
 
+#[derive(Serialize, Deserialize)]
+struct CoreInfo {
```
Comment (atsushi421) [inline lib/src/output_log.rs:25]:
```suggestion
struct ProcessorInfo {
```

File: lib/src/output_log.rs [hunk 2]
```diff
@@ -32,7 +38,29 @@ pub fn create_yaml_file(folder_path: &str, file_name: &str) -> String {
...
+        }
+    } else {
+        eprintln!("Failed to open file: {}", file_path);
+    }
+}
+
+pub fn dump_number_of_cores_to_yaml(file_path: &str, number_of_cores: usize) {
+    let number_of_cores_info = CoreInfo { number_of_cores };
+    let yaml = serde_yaml::to_string(&number_of_cores_info)
+        .expect("Failed to serialize DAGSetInfo to YAML");
```
Comment (atsushi421) [inline lib/src/output_log.rs:59]: nits

File: lib/src/output_log.rs [hunk 3]
```diff
@@ -32,7 +38,29 @@ pub fn create_yaml_file(folder_path: &str, file_name: &str) -> String {
...
+    {
+        if let Err(err) = file.write_all(info.as_bytes()) {
+            eprintln!("Failed to write to file: {}", err);
+        }
+    } else {
+        eprintln!("Failed to open file: {}", file_path);
+    }
+}
+
+pub fn dump_number_of_cores_to_yaml(file_path: &str, number_of_cores: usize) {
```
Comment (atsushi421) [inline lib/src/output_log.rs:56]:
```suggestion
pub fn dump_processor_info_to_yaml(file_path: &str, number_of_cores: usize) {
```

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        } => ResultInfo {
+            result: result.clone(),
+        },
+        FederateResult::Schedulable {
+            high_dedicated_cores: _,
+            low_dedicated_cores: _,
+        } => ResultInfo {
+            result: result.clone(),
+        },
+    };
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:25]:
```suggestion
    let result_info = ResultInfo { result };
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: lib/src/output_log.rs (hunk 1 above)
Comment (kobayu858) [reply lib/src/output_log.rs]: CoreInfo -> ProcessorInfo

File: lib/src/output_log.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/output_log.rs:59]: CoreInfo -> ProcessorInfo

Comment (kobayu858) [reply lib/src/output_log.rs]: DAGSetInfo -> ProcessorInfo

File: lib/src/output_log.rs (hunk 3 above)
Comment (kobayu858) [reply lib/src/output_log.rs]: dump_number_of_cores_to_yaml -> dump_processor_info_to_yaml

File: 2014_RTSS_federated_original/src/outputs_result.rs (hunk 4 above)
Comment (kobayu858) [reply 2014_RTSS_federated_original/src/outputs_result.rs]: accept your suggestion

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        let mut dag = Graph::<NodeData, f32>::new();
+        let mut params = HashMap::new();
+        params.insert("execution_time".to_owned(), 20.0);
+        params.insert("period".to_owned(), 10.0);
+        dag.add_node(NodeData { id: 0, params });
+        dag
+    }
+
+    #[test]
+    fn test_dump_federated_info_normal() {
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:77]: Rename

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                high_dedicated_cores: 6,
+                low_dedicated_cores: 34,
+            }
+        );
+
+        remove_file(file_path).unwrap();
+    }
+
+    #[test]
+    fn test_dump_federated_lack_cores_for_high_tasks() {
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:103]: Rename

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                reason: (String::from("Insufficient number of cores for high-utilization tasks.")),
+                insufficient_cores: 2
+            }
+        );
+
+        remove_file(file_path).unwrap();
+    }
+
+    #[test]
+    fn test_dump_federated_lack_cores_for_low_tasks() {
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:129]: Rename

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                reason: (String::from("Insufficient number of cores for low-utilization tasks.")),
+                insufficient_cores: 2
+            }
+        );
+
+        remove_file(file_path).unwrap();
+    }
+
+    #[test]
+    fn test_dump_federated_unsuited_tasks() {
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:155]: Rename

File: 2014_RTSS_federated_original/src/outputs_result.rs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+use crate::federated::FederateResult;
+
+#[derive(Serialize, Deserialize)]
+struct ResultInfo<FederateResult> {
+    result: FederateResult,
+}
+
+pub fn dump_federate_result_to_file(file_path: &str, result: FederateResult) {
+    let result_info = ResultInfo { result };
+
```
Comment (atsushi421) [inline 2014_RTSS_federated_original/src/outputs_result.rs:13]:
```suggestion
```

File: lib/src/output_log.rs [hunk 10]
```diff
@@ -109,4 +141,18 @@ mod tests {
 
         remove_file(file_path).unwrap();
     }
+
+    #[test]
+    fn test_dump_number_of_cores_info_to_yaml() {
```
Comment (atsushi421) [inline lib/src/output_log.rs:146]: Rename

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_RTSS_federated_original/src/outputs_result.rs (hunk 6 above)
Comment (kobayu858) [reply 2014_RTSS_federated_original/src/outputs_result.rs]: rename

**Review by atsushi421 — APPROVED** [review]
LGTM!

**Pull request merged**